### PR TITLE
`schemadiff`: supporting textual diff

### PR DESCRIPTION
**Pull request by shlomi-noach**
## Description

In this PR `schemadiff` is able to generate textual diffs, nuanced to diff hints and specific logic of `schemadiff`.

Consider the following example. Two schemas are diffed, and the logical diff is:
```sql
DROP TABLE `t1`
ALTER TABLE `t2` MODIFY COLUMN `id` bigint
CREATE TABLE `t4` (
	`id` int,
	PRIMARY KEY (`id`)
)
```

On top of this logical diff, `schemadiff` will now also be able to generate this textual diff:

```diff
-CREATE TABLE `t1` (
-	`id` int,
-	PRIMARY KEY (`id`)
-)
 CREATE TABLE `t2` (
-	`id` int,
+	`id` bigint,
 	PRIMARY KEY (`id`)
 )
+CREATE TABLE `t4` (
+	`id` int,
+	PRIMARY KEY (`id`)
+)
```

The above is a simple example. What sets `schemadiff` apart is that the textual diff only shows what `schemadiff` thinks are ntoeworthy changes. For example, `schemadiff` ignores index ordering: it considers two tables to be identical if all but order of indexes is the same. The standard textual diff of such two tables will show the indexes as `+` and `-` annotated entries, but the `schemdiff` textual diff will show no changes. Likewise, based on diff hints, constraint names can be ignored, etc.

It also takes care of trailing last comma issues, so common with e.g. `git diff`. For example, in the below:
```sql
CREATE TABLE `t4` (
	`id` int,
	PRIMARY KEY (`id`)
)
```
```sql
CREATE TABLE `t4` (
	`id` int,
	PRIMARY KEY (`id`),
	KEY `another_idx` (`id`)
)
```
A standard textual diff will capture the `PRIMARY KEY` line, due to the addition of the trailing comma. `schemadiff`, however, outputs:
```diff
 CREATE TABLE `t4` (
 	`id` int,
 	PRIMARY KEY (`id`),
+	KEY `another_idx` (`id`)
 )
```

Implementation-wise, a `DiffEntity` now has an `Annotated()` function, which returns 3 annotated diff texts:

1+2: The split from & to annotated forms.
3: The unified form (as shown in the above example).

## Related Issue(s)

Fixes https://github.com/vitessio/vitess/issues/15387

## Checklist

-   [ ] "Backport to:" labels have been added if this change should be back-ported to release branches
-   [ ] If this change is to be back-ported to previous releases, a justification is included in the PR description
-   [ ] Tests were added or are not required
-   [ ] Did the new or modified tests pass consistently locally and on CI?
-   [ ] Documentation was added or is not required

### CODE REVIEW

**Comment by shlomi-noach** [comment]
Will probably revert to `b23bb2dca65ea9de007290cd320ccbc16ac1e306` and undo the HTML changes, as they're too naive. They would for example mark a column name that happens to also be a keyword.

**Comment by shlomi-noach** [comment]
HTML change reverted via force push.

File: go/vt/schemadiff/table.go [hunk 1]
```diff
@@ -1234,13 +1266,15 @@ func (c *CreateTableEntity) diffPartitions(alterTable *sqlparser.AlterTable,
 	case t1Partitions == nil:
 		// add partitioning
 		alterTable.PartitionOption = t2Partitions
+		annotations.MarkAdded(sqlparser.CanonicalString(t2Partitions))
```
Comment (mattlord) [inline go/vt/schemadiff/table.go:1269]: Any reason not to add a test case for this in the unit test?

File: go/vt/schemadiff/table.go [hunk 2]
```diff
@@ -1111,37 +1135,41 @@ func (c *CreateTableEntity) diffOptions(alterTable *sqlparser.AlterTable,
...
-							alterTableOptions = append(alterTableOptions, t2Option)
+							modifyTableOption(t1Option, t2Option)
 						}
 					case AutoIncrementIgnore:
 						// do not apply
 					}
 				default:
 					// Apply the new options
-					alterTableOptions = append(alterTableOptions, t2Option)
+					modifyTableOption(t1Option, t2Option)
```
Comment (mattlord) [inline go/vt/schemadiff/table.go:1145]: Seems like it's probably relatively easy to cover the default in the unit test too?

File: go/vt/schemadiff/table.go [hunk 3]
```diff
@@ -1087,10 +1111,10 @@ func (c *CreateTableEntity) diffOptions(alterTable *sqlparser.AlterTable,
 				case "CHARSET", "COLLATE":
 					switch hints.TableCharsetCollateStrategy {
 					case TableCharsetCollateStrict:
-						alterTableOptions = append(alterTableOptions, t2Option)
+						modifyTableOption(t1Option, t2Option)
 					case TableCharsetCollateIgnoreEmpty:
 						if t1Option.String != "" && t2Option.String != "" {
-							alterTableOptions = append(alterTableOptions, t2Option)
+							modifyTableOption(t1Option, t2Option)
```
Comment (mattlord) [inline go/vt/schemadiff/table.go:1117]: Looks like this isn't covered by the unit test.

**Review by mattlord — APPROVED** [review]
Nice! Only a few very minor comments about potential additional unit test cases. If they're relatively easy to add, probably worth it. I'll leave that up to you though.

File: go/vt/schemadiff/table.go (hunk 1 above)
Comment (shlomi-noach) [reply go/vt/schemadiff/table.go:1269]: Added.

File: go/vt/schemadiff/table.go (hunk 2 above)
Comment (shlomi-noach) [reply go/vt/schemadiff/table.go:1145]: Added.

File: go/vt/schemadiff/table.go (hunk 3 above)
Comment (shlomi-noach) [reply go/vt/schemadiff/table.go:1117]: Added,

**Review by notfelineit — APPROVED** [review]

**Pull request merged**